File: frontend/src/app/features/pedidos/domain/entities/detalle_pedido.ts
/**
 * Entidad de dominio que representa una línea de detalle de un pedido.
 * @remarks Incluye el importe ya calculado (`importeLinea`) y el descuento aplicado.
 */
export class DetallePedido {
  private readonly _detallePedidoID: number;
  private readonly _pedidoID: number;
  private readonly _productoID: number;
  private readonly _productoNombre: string;
  private readonly _codigoProducto: string;
  private readonly _cantidad: number;
  private readonly _precioUnitario: number;
  private readonly _descuento: number;
  private readonly _importeLinea: number;
  private readonly _observaciones: string | null;

  constructor(
    detallePedidoID: number,
    pedidoID: number,
    productoID: number,
    productoNombre: string,
    codigoProducto: string,
    cantidad: number,
    precioUnitario: number,
    descuento: number,
    importeLinea: number,
    observaciones: string | null = null
  ) {
    this._detallePedidoID = detallePedidoID;
    this._pedidoID = pedidoID;
    this._productoID = productoID;
    this._productoNombre = productoNombre;
    this._codigoProducto = codigoProducto;
    this._cantidad = cantidad;
    this._precioUnitario = precioUnitario;
    this._descuento = descuento;
    this._importeLinea = importeLinea;
    this._observaciones = observaciones;
  }

  /** @returns Identificador único de la línea de detalle. */
  public get detallePedidoID(): number {
    return this._detallePedidoID;
  }

  /** @returns Identificador del pedido al que pertenece esta línea. */
  public get pedidoID(): number {
    return this._pedidoID;
  }

  /** @returns Identificador del producto solicitado. */
  public get productoID(): number {
    return this._productoID;
  }

  /** @returns Nombre del producto en el momento de crear la línea. */
  public get productoNombre(): string {
    return this._productoNombre;
  }

  /** @returns Código interno del producto. */
  public get codigoProducto(): string {
    return this._codigoProducto;
  }

  /** @returns Cantidad de unidades solicitadas. */
  public get cantidad(): number {
    return this._cantidad;
  }

  /** @returns Precio unitario acordado con el proveedor. */
  public get precioUnitario(): number {
    return this._precioUnitario;
  }

  /** @returns Porcentaje de descuento aplicado (0–100). */
  public get descuento(): number {
    return this._descuento;
  }

  /** @returns Importe total de la línea tras aplicar descuento. */
  public get importeLinea(): number {
    return this._importeLinea;
  }

  /** @returns Observaciones de la línea o `null` si no las hay. */
  public get observaciones(): string | null {
    return this._observaciones;
  }
}

File: backend/ERP.Backend/Domain/Entities/DetallePedido.cs
using System.ComponentModel.DataAnnotations;

namespace Domain.Entities
{
    /// <summary>
    /// Entidad de dominio que representa una línea de detalle dentro de un pedido a proveedor.
    /// El campo <see cref="ImporteLinea"/> es una columna calculada y almacenada en base de datos.
    /// </summary>
    public class DetallePedido
    {
        #region atributos
        private int _detallePedidoID;
        private int _pedidoID;
        private int _productoID;
        private int _cantidad;
        private decimal _precioUnitario;
        private decimal _descuento;
        private bool _activo;
        #endregion

        #region constructores
        /// <summary>
        /// Constructor requerido por el ORM / serializadores.
        /// No debe utilizarse directamente en lógica de dominio.
        /// </summary>
        public DetallePedido() { }

        /// <summary>
        /// Crea una línea de detalle de pedido con todos sus valores iniciales.
        /// </summary>
        /// <param name="detallePedidoID">Identificador único del detalle.</param>
        /// <param name="pedidoID">Identificador del pedido al que pertenece.</param>
        /// <param name="productoID">Identificador del producto solicitado.</param>
        /// <param name="cantidad">Cantidad de unidades solicitadas.</param>
        /// <param name="precioUnitario">Precio unitario acordado.</param>
        /// <param name="descuento">Porcentaje de descuento aplicado (0–100).</param>
        /// <param name="activo">Indica si la línea está activa.</param>
        public DetallePedido(int detallePedidoID, int pedidoID, int productoID,
            int cantidad, decimal precioUnitario, decimal descuento, bool activo)
        {
            _detallePedidoID = detallePedidoID;
            _pedidoID = pedidoID;
            _productoID = productoID;
            _cantidad = cantidad;
            _precioUnitario = precioUnitario;
            _descuento = descuento;
            _activo = activo;
        }
        #endregion

        #region propiedades
        /// <summary>Identificador único del detalle (clave primaria).</summary>
        [Key]
        public int DetallePedidoID { get => _detallePedidoID; set => _detallePedidoID = value; }

        /// <summary>Identificador del pedido al que pertenece esta línea.</summary>
        [Required]
        public int PedidoID { get => _pedidoID; set => _pedidoID = value; }

        /// <summary>Identificador del producto solicitado.</summary>
        [Required]
        public int ProductoID { get => _productoID; set => _productoID = value; }

        /// <summary>Cantidad de unidades solicitadas. Debe ser mayor que 0.</summary>
        [Required]
        [Range(1, int.MaxValue, ErrorMessage = "La cantidad debe ser mayor que 0")]
        public int Cantidad { get => _cantidad; set => _cantidad = value; }

        /// <summary>Precio unitario acordado. Debe ser mayor que 0.</summary>
        [Required]
        [Range(0.01, double.MaxValue, ErrorMessage = "El precio unitario debe ser mayor que 0")]
        public decimal PrecioUnitario { get => _precioUnitario; set => _precioUnitario = value; }

        /// <summary>Porcentaje de descuento aplicado (0–100).</summary>
        [Required]
        [Range(0, 100, ErrorMessage = "El descuento debe estar entre 0 y 100")]
        public decimal Descuento { get => _descuento; set => _descuento = value; }

        /// <summary>
        /// Importe total de la línea calculado en BD como
        /// <c>Cantidad × PrecioUnitario × (1 - Descuento / 100)</c>. Solo lectura.
        /// </summary>
        public decimal ImporteLinea { get; private set; }

        /// <summary>Indica si la línea está activa. Las líneas inactivas se excluyen de los cálculos.</summary>
        [Required]
        public bool Activo { get => _activo; set => _activo = value; }


        /// <summary>Referencia de navegación al pedido padre.</summary>
        public Pedido? Pedido { get; set; }

        /// <summary>Referencia de navegación al producto de esta línea.</summary>
        public Producto? Producto { get; set; }
        #endregion
    }
}